rep-2012.rst

```diff
@@ -0,0 +1,466 @@
+Or, we could post-process recorded requests and responses to validate their content.
+
+There are additional features that could leverage this proposal, such as:
+
+- Playback of recorded services (for example, from a rosbag [1]_)
+- Introspection of ROS actions, which are built on services
+- Validation of a live ROS system by referencing a recording from a previous session
+
+Though this proposal focuses on the core feature of introspecting requests and responses, we want to keep the design flexible so additional features like those listed above can be implemented in the future.
+
+
+Specification
+=============
+
+Publishing Service Events
+-------------------------
+
+Whenever a request or response is sent or received, a *service event* message will be published to a topic.
+Services are responsible for publishing a message when they receive a request and when they send a response.
+Likewise, clients are responsible for publishing a message when they send a request and when they receive a response.
+Therefore, we have a total of four possible events:
+
+:Request Sent:
+  Emitted from a client after sending a request to a service.
+:Request Received:
+  Emitted from a service after receiving a request from a client.
+:Response Sent:
+  Emitted from a service after sending a response to a client.
+:Response Received:
+  Emitted from a client after receiving a response from a service.
+
+Request event messages shall be published to the hidden topic ``/SERVICE_NAME/_request_event``, where ``SERVICE_NAME`` is the fully-qualified name of the service.
+Similarly, response event messages shall be published to the hidden topic ``/SERVICE_NAME/_response_event``.
+Note that this implies that services must have unique names.
+
+By publishing service event messages to predetermined topics, tools and libraries are able to subscribe to these topics to inspect the flow of data between services.
+
+Service Event Definitions
+-------------------------
+
+For each service definition, ``my/srv/Foo``, we define two new ROS message types using the ROS IDL specification [2]_:
+
+:my/msg/Foo_RequestEvent:
+  This type communicates a request sent or request received event.
+
+  .. code-block::
+
+     # Indicates this request event was emitted from a client
+     uint8 CLIENT_SENT      = 0
+
+     # Indicates this request event was emitted from a service
+     uint8 SERVICE_RECEIVED = 1
+
+     # Whether this is a CLIENT_SENT or SERVICE_RECEIVED event.
+     uint8 request_type
+
+     # Timestamp for when the event occurred (sent or received time)
+     builtin_interfaces/msg/Time stamp
+
+     # Unique identifier for the client that sent the request
+     # Note, this is only unique for the current session
+     unique_identifier_msgs/msg/UUID client_id
+
+     # Sequence number for the request
+     # Combined with the client ID, this creates a unique ID for the request
+     int64 sequence_number
```

Comment:
To aid in the implementation, I think this set of members should be grouped into an aggregate message, e.g. `rcl_interfaces/msg/ServiceEventInfo.msg`

Reply:
We already have a sequence number in `std_msgs/Header` and we're overloading the terminology. One suggestion is to perhaps call this an `interaction_id` and consider making it a randomly-generated UUID. I say this because I worry that there's some weird edge case involving stitching bags together, each of which has queries on the same service that if started at `sequence_number=0` would destructively interfere with each other. Ultimately, this value is probably going to be shoved into a hash table to quickly associate events into interactions, so we should consider designing it in a way that it will be used.

Reply:
This sequence number is specifically coming from the middleware, it is not intended to be something new introduced by the service introspection feature (just exposing the existing info of a service to a topic). 

The value is copied from the sequence number associated with the request ID:

https://github.com/ros2/rmw/blob/2259c3f6f3de8c479a9d5c74cdcd03d0010413cd/rmw/include/rmw/types.h#L361-L362

Reply:
FWIW, the sequence number alone is not a unique identifier for the request. It should be combined with the Client GUID to make a unique identifier for the session. And, there is no guarantee it will be unique across different sessions i.e. there's no guarantee the IDs will be the same (or different) the next time you launch your ROS system.